src/vsphere_cpi/lib/cloud/vsphere/nsxt_policy_provider.rb

```diff
@@ -0,0 +1,360 @@
+      Bosh::Retryable.new(
+          tries: @max_tries,
+          sleep: ->(try_count, retry_exception) { @sleep_time },
+          on: [UnrealizedResource]
+      ).retryer do |i|
+        logger.info("Polling (try:##{i}) to check if Entity with path: #{intent_path} is realized or not")
+        result = policy_realization_api.list_realized_entities(intent_path)
+
+        # In case policy API has not started realizing an entity. the results should be nil then as count is 0
+        raise UnrealizedResource.new(intent_path) if result.result_count == 0
+        # A single policy resource might need multiple manager realizations to happen
+        # before it could be use. Here we are checking that all the
+        # GenericPolicyRealizedResource objects (the results) have been realized
+        # and raise an error if any of the realization object is in ERROR state.
+        realized = result.results.all? do |res|
+          if res.state == 'ERROR'
+            raise ErrorRealizingResource.new("Error realizing entity #{res} for intent path: #{intent_path}")
+          end
+          res.state == 'REALIZED'
+        end
+        # If none has been realized yet, raise error
+        raise UnrealizedResource.new(intent_path) unless realized
+        logger.info("Context VM: #{vm.cid unless vm.nil?} Entity: #{intent_path} realized.")
+      end
+    end
+
+    def retrieve_group(group_id:)
+      logger.info("Searching for Policy Group with group id: #{group_id}")
+      begin
+        group = policy_group_api.read_group_for_domain(DEFAULT_NSXT_POLICY_DOMAIN, group_id)
+      rescue => e
+        logger.info("Failed to find Policy Group: #{group_id} with error #{e.inspect}")
+        raise e
+      end
+      logger.info("Found Policy Group: #{group.id}")
+      group
+    end
+
+    def retrieve_segment_port(seg:, port:)
+      logger.info("Searching for segment port: #{port} on segment: #{seg}")
+      begin
+        segment_port = policy_segment_port_api.get_infra_segment_port(seg, port)
+      rescue => e
+        logger.info("Failed to find port: #{port} on segment: #{seg}")
+        raise e
+      end
+      segment_port
+    end
+
+    def retrieve_all_with_pagination(object_type:)
+      logger.info("Gathering all #{object_type}")
+      objects = []
+      result_list = yield
+      objects.push(*result_list.results)
+      until result_list.cursor.nil?
+        result_list = yield(result_list.cursor)
+        objects.push(*result_list.results)
+      end
+      logger.info("Found #{objects.size} number of #{object_type}")
+      objects
+    end
+
+    def retrieve_all_groups
+      retrieve_all_with_pagination(object_type: 'Policy Group') do |cursor=nil|
+        unless cursor.nil?
+          policy_group_api.list_group_for_domain(DEFAULT_NSXT_POLICY_DOMAIN, cursor: cursor)
+        else
+          policy_group_api.list_group_for_domain(DEFAULT_NSXT_POLICY_DOMAIN)
+        end
+      end
+    end
+
+    def retry_on_conflict(log_str)
```

Comment:
@nehagjain15 @ktchen14 : Need help to express this better. Want to avoid API Error due to high requests/second (`>100`). Suggestions to code it better?